cumulusci/tasks/bulkdata/tests/test_upsert.py

```diff
@@ -0,0 +1,400 @@
+        <systemModstamp>2022-03-02T04:14:44.000Z</systemModstamp>
+
+        <numberRecordsProcessed>0</numberRecordsProcessed>
+        <numberRecordsFailed>0</numberRecordsFailed>
+
+        <totalProcessingTime>0</totalProcessingTime>
+        <apiActiveProcessingTime>0</apiActiveProcessingTime>
+
+        <apexProcessingTime>0</apexProcessingTime>
+        </batchInfo>""",
+        )
+
+        responses.add(
+            method="GET",
+            url=f"https://orgname.my.salesforce.com/services/async/{current_sf_version}/job/750P0000005mX3LIAU/batch/751P0000006fIUAIA2/result",
+            status=200,
+            body="""""Id","Success","Created","Error"
+        "003P000001Y5yvdIAB","true","false",""
+        "003P000001Y5ywMIAR","true","false",""
+        "003P000001Y5ywNIAR","true","false",""
+        "003P000001Y5ywOIAR","true","false","" """,
+        )
+
+        task = create_task(
+            LoadData,
+            {
+                "sql_path": cumulusci_test_repo_root
+                / "datasets/upsert/upsert_example_2.sql",
+                "mapping": cumulusci_test_repo_root
+                / "datasets/upsert/upsert_mapping_bulk.yml",
+                "ignore_row_errors": True,
+                "set_recently_viewed": False,
+            },
+        )
+
+        with mock.patch.object(task.logger, "debug"):
+            ret = task()
+            assert ret == {
+                "step_results": {
+                    "Insert Accounts": {
+                        "sobject": "Account",
+                        "record_type": None,
+                        "status": DataOperationStatus.SUCCESS,
+                        "job_errors": [],
+                        "records_processed": 0,
+                        "total_row_errors": 0,
+                    },
+                    "Insert Contacts": {
+                        "sobject": "Contact",
+                        "record_type": None,
+                        "status": DataOperationStatus.SUCCESS,
+                        "job_errors": [],
+                        "records_processed": 0,
+                        "total_row_errors": 0,
+                    },
+                }
+            }, ret
+            relevant_debug_statement = look_for_operation_creation_debug_statement(
+                task.logger.debug.mock_calls
+            )
+            assert relevant_debug_statement == str(
+                DataApi.BULK
+            ), relevant_debug_statement
+
+
+def look_for_operation_creation_debug_statement(mock_calls):
+    relevant_debug_statements = (
+        look_for_operation_creation_debug_statement_for_string(call.args[0])
+        for call in mock_calls
+    )
+
+    return next(stmt for stmt in relevant_debug_statements if stmt)
+
+
+DEBUG_MATCHER = re.compile("Creating (.*) Operation for (.*) using (.*)")
+
+
+def look_for_operation_creation_debug_statement_for_string(s):
+    match = DEBUG_MATCHER.match(s)
+    if match:
+        return match[3]
```

Comment:
Yes, or some form of callback mechanism would also work.

Actually, the debug thing was a bit cleaner in a previous iteration. One could theoretically do:

logger.debug("Message %s %s", A, B) 

And then by mocking logger.debug, you don't need to parse. This worked except in the context of robot, where we inject a logger which is not compatible with debug-with-VARARGS.

Using debug in this way might hit the 80/20 point and give us some additional debug logging that would be generally useful in debugging problems. Fixing Robot to accept it is probably not too hard.

Reply:
I think you're saying the same thing I was trying to day. The reason I'm logging the formatted string is because the Robot logger does not support the "structured logging".